### PR TITLE
KeepWorking badges on header bubbles and script overview

### PR DESCRIPTION
**Pull request by maureensturgeon**
The changes in this PR add "keep working" badges to the progress bubbles in the header, script overview pages and to the progress bubbles on the sublevel and bonus level cards. These badges should show if the teacher has given `keepWorking` feedback to the student for the level.

There's a lot to review here, if it's too much to give a good review let me know and I'll figure out a way to split it up into multiple PRs.

Screen Shot 2021-07-01 at 4 22 40 PM

Screen Shot 2021-07-02 at 1 31 53 PM

## Links

- jira ticket: [LP-1900](https://codedotorg.atlassian.net/browse/LP-1900)

## Testing story
Tested locally for several levels: regular, bubble choice, assessment, also added unit tests.

## PR Checklist:

- [ ] Tests provide adequate coverage
- [ ] Privacy and Security impacts have been assessed
- [ ] Code is well-commented
- [ ] New features are translatable or updates will not break translations
- [ ] Relevant documentation has been added or updated
- [ ] User impact is well-understood and desirable
- [ ] Pull Request is labeled appropriately
- [ ] Follow-up work items (including potential tech debt) are tracked and linked

### CODE REVIEW

File: apps/src/templates/progress/ProgressPill.jsx [hunk 1]
```diff
@@ -29,53 +34,83 @@ class ProgressPill extends React.Component {
...
 
+  getUrl() {
+    const {
+      levels,
+      disabled,
+      selectedSectionId,
+      onSingleLevelClick
+    } = this.props;
+
+    if (levels.length > 1 || disabled || onSingleLevelClick) {
```
Comment (tim-dot-org) [inline apps/src/templates/progress/ProgressPill.jsx]: could you extract this or add a comment to say what its intended to capture

File: dashboard/app/controllers/script_levels_controller.rb [hunk 2]
```diff
@@ -108,6 +105,9 @@ def show
       return
     end
 
+    # will be true if the user is in any unarchived section where tts autoplay is enabled
+    @tts_autoplay_enabled = current_user&.sections_as_student&.where({hidden: false})&.map(&:tts_autoplay_enabled)&.reduce(false, :|)
```
Comment (tim-dot-org) [inline dashboard/app/controllers/script_levels_controller.rb:109]: was this change intended? seems unrelated

File: dashboard/app/helpers/users_helper.rb [hunk 3]
```diff
@@ -203,30 +235,40 @@ def script_progress_for_users(users, script)
...
         if level.is_a?(BubbleChoice) # we have a parent level
           bubble_choice_progress = get_bubble_choice_progress(
-            level, user, user_levels_by_level, sl, paired_user_levels, include_timestamp
+            level, user, user_levels_by_level, teacher_feedback_by_level, sl, paired_user_levels, include_timestamp
           )
-          if bubble_choice_progress
+          if bubble_choice_progress.present?
             progress.merge!(bubble_choice_progress.compact)
+
+            sum_time_spent = bubble_choice_progress&.values&.reduce(0) do |sum, sublevel_progress|
```
Comment (tim-dot-org) [inline dashboard/app/helpers/users_helper.rb]: i think these safe accesses could cause `sum_time_spent` to resolve to nil, and then cause an error at `sum_time_spent > 0` later on, maybe default to 0 or add another check?

**Review by tim-dot-org — APPROVED** [review]
love all the tidying and the tests definitely helped me understand things more!
dunno if i have enough context to give full approval, but looks good to me apart from a couple things

File: dashboard/app/controllers/script_levels_controller.rb (hunk 2 above)
Comment (maureensturgeon) [reply dashboard/app/controllers/script_levels_controller.rb:109]: Oh I just moved that down below the redirect so that we wouldn't do an unnecessary query just to redirect afterwards! It was just a little optimization I noticed when I was looking through the code.

File: dashboard/app/helpers/users_helper.rb (hunk 3 above)
Comment (maureensturgeon) [reply dashboard/app/helpers/users_helper.rb]: Definitely worth checking, will update!

Comment (maureensturgeon) [reply dashboard/app/helpers/users_helper.rb]: So I realized that sum_time_spent is within a block checking `bubble_choice_progress.present?`, so I'll remove those safe operators, it'll never be nil

File: apps/test/unit/code-studio/components/SublevelCardTest.js [hunk 4]
```diff
@@ -1,98 +1,87 @@
...
       wrapper.find('.sublevel-card-title-uitest').text()
     );
     assert.equal(
-      LESSON_EXTRAS_DEFAULT_PROPS.sublevel.description,
+      DEFAULT_SUBLEVEL.description,
       wrapper.find('.sublevel-card-description-uitest').text()
     );
   });
+
+  it('mapps sublevel keys to camelcase before passing to ProgressBubble', () => {
```
Comment (cforkish) [inline apps/test/unit/code-studio/components/SublevelCardTest.js]: NIT: typo

File: dashboard/app/helpers/users_helper.rb [hunk 5]
```diff
@@ -203,30 +235,40 @@ def script_progress_for_users(users, script)
...
     progress = {}
     script.script_levels.each do |sl|
       sl.level_ids.each do |level_id|
         level = Level.cache_find(level_id)
+        level_for_progress = level.get_level_for_progress(user, script)
+        ul = user_levels_by_level.try(:[], level_for_progress.id)
+
+        # For contained levels, progress is stored with the contained level id but feedback is stored with
+        # the level id, which is why we need a different level id to get feedback for contained levels.
+        level_id_for_feedback = level.contained_levels.any? ? level.id : level_for_progress.id
```
Comment (cforkish) [inline dashboard/app/helpers/users_helper.rb:251]: could we avoid an extra query for non-contained levels if we move line 246 into an`if level.contained_levels.any?` block?

File: apps/src/code-studio/progressRedux.js [hunk 6]
```diff
@@ -611,7 +613,8 @@ const levelWithProgress = (
     status: status,
     isCurrentLevel: isCurrent,
     paired: levelPairing[level.activeId],
-    isLocked: locked
+    isLocked: locked,
+    teacherFeedbackReviewState: teacherFeedbackReviewState
```
Comment (cforkish) [inline apps/src/code-studio/progressRedux.js:621]: could be nice to add this to `LevelWithProgressType` in `progressTypes.js`

**Review by cforkish — APPROVED** [review]
Nice work! This is big, but very readable. :)

A few minor comments, but nothing blocking.

File: dashboard/app/helpers/users_helper.rb (hunk 5 above)
Comment (maureensturgeon) [reply dashboard/app/helpers/users_helper.rb:251]: `get_level_for_progress` will get the contained level OR the sublevel, it's not just for contained levels unfortunately.

File: apps/src/code-studio/progressRedux.js (hunk 6 above)
Comment (maureensturgeon) [reply apps/src/code-studio/progressRedux.js:621]: Ah right, that's something that gets added in https://github.com/code-dot-org/code-dot-org/pull/41434 but I could add it here.

Comment (maureensturgeon) [reply apps/src/code-studio/progressRedux.js:621]: Added!
